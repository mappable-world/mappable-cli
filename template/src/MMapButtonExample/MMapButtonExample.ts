import type {MMapControlButton, MMapControlButtonProps} from '@mappable-world/mappable-types';

interface MMapButtonExampleProps extends MMapControlButtonProps {
    className?: string;
}

export class MMapButtonExample extends mappable.MMapComplexEntity<MMapButtonExampleProps> {
    private _button?: MMapControlButton;
    private _element: HTMLElement = document.createElement('div');

    constructor(props: MMapButtonExampleProps) {
        super(props);
        this._element.classList.add(this._props.className);
    }

    protected _onAttach() {
        this._button = new mappable.MMapControlButton({...this._props, element: this._element});
        this.addChild(this._button);
    }

    protected _onUpdate(props: Partial<MMapButtonExampleProps>, oldProps: MMapButtonExampleProps) {
        this._button.update(props);

        if (props.className !== undefined) {
            this._element.classList.remove(oldProps.className);
            this._element.classList.add(props.className);
        }
    }

    protected _onDetach() {
        this.removeChild(this._button);
    }
}
