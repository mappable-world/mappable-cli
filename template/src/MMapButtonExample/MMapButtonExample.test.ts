import type {LngLatBounds} from '@mappable-world/mappable-types/common/types';
import {MMap} from '@mappable-world/mappable-types';
import {MMapButtonExample} from './MMapButtonExample';

describe('MMap smoke test', () => {
    const BOUNDS: LngLatBounds = [
        [54.58311, 25.9985],
        [56.30248, 24.47889]
    ];

    const LOCATION = {bounds: BOUNDS};

    let container: HTMLElement, map: MMap;
    beforeEach(() => {
        container = document.createElement('div');
        Object.assign(container.style, {width: `643px`, height: `856px`});
        document.body.appendChild(container);
        map = new mappable.MMap(container, {location: LOCATION}, [new mappable.MMapDefaultSchemeLayer({})]);
    });

    afterEach(() => {
        map.destroy();
        container.remove();
    });

    it('should make map', () => {
        map.addChild(
            new mappable.MMapControls({position: 'bottom'}, [
                new MMapButtonExample({
                    text: 'Some text',
                    className: 'user-class',
                    onClick: () => {
                        console.log('Click!');
                    }
                })
            ])
        );

        const tree = domToJson(map.container);
        expect(tree).toMatchSnapshot();
    });
});
