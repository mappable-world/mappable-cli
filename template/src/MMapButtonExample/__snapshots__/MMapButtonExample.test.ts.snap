// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MMap smoke test should make map 1`] = `
{
  "attributes": {
    "class": "mappable--map-container",
  },
  "children": [
    {
      "attributes": {
        "class": "mappable--map",
        "style": "width: 643px; height: 856px;",
      },
      "children": [
        {
          "attributes": {
            "class": "mappable--main-engine-container",
          },
          "children": [
            {
              "attributes": {},
              "children": [
                {
                  "attributes": {
                    "class": "mappable--layer mappable--tile-layer",
                    "style": "perspective: 1597.3177456394797px; opacity: 1;",
                  },
                  "children": [
                    {
                      "attributes": {
                        "class": "mappable--tile-layer__container",
                        "data-z": "9",
                        "style": "transform: 
            rotateX(0.00000rad)
            translate3d(0px, 0px, 0px)
            rotateZ(0.00000rad) translate3d(0px, 0px, 0px)
        ;",
                      },
                      "nodeName": "mappable",
                    },
                  ],
                  "nodeName": "mappable",
                },
              ],
              "nodeName": "mappable",
            },
          ],
          "nodeName": "mappable",
        },
        {
          "attributes": {
            "class": "mappable--top-engine-container",
          },
          "children": [
            {
              "attributes": {},
              "nodeName": "mappable",
            },
          ],
          "nodeName": "mappable",
        },
      ],
      "nodeName": "mappable",
    },
    {
      "attributes": {
        "class": "mappable--controls mappable--controls_bottom mappable--controls_center mappable--controls_horizontal",
      },
      "children": [
        {
          "attributes": {
            "class": "mappable--control mappable--control__background",
          },
          "children": [
            {
              "attributes": {
                "class": "mappable--button mappable--control-button",
                "type": "button",
              },
              "children": [
                "Some text",
              ],
              "nodeName": "button",
            },
          ],
          "nodeName": "mappable",
        },
      ],
      "nodeName": "mappable",
    },
    {
      "attributes": {
        "class": "mappable--map-copyrights mappable--map-copyrights_bottom mappable--map-copyrights_right",
      },
      "children": [
        {
          "attributes": {
            "class": "mappable--map-copyrights__container",
          },
          "children": [
            {
              "attributes": {
                "class": "mappable--map-copyrights__text",
              },
              "nodeName": "span",
            },
            {
              "attributes": {
                "class": "mappable--map-copyrights__user-agreements",
                "href": "https://mappable.world/legal/tos",
                "target": "_blank",
              },
              "children": [
                "User Agreement",
              ],
              "nodeName": "a",
            },
          ],
          "nodeName": "div",
        },
        {
          "attributes": {
            "class": "mappable--icon-logo-en-black mappable--map-copyrights__logo",
            "style": "width: 80px; height: 20px;",
          },
          "nodeName": "span",
        },
        {
          "attributes": {
            "class": "mappable--map-copyrights__scale",
          },
          "nodeName": "div",
        },
      ],
      "nodeName": "div",
    },
  ],
  "nodeName": "mappable",
}
`;
